src/main/java/com/rocketden/main/service/RoomService.java

```diff
@@ -47,6 +47,10 @@ public RoomDto joinRoom(String roomId, JoinRoomRequest request) {
             throw new ApiException(RoomError.NOT_FOUND);
         }
 
+        if (room.isFull()) {
+            throw new ApiException(RoomError.ALREADY_FULL);
+        }
```

Comment:
I wonder if two new users attempt to join simultaneously, if there is any chance that the `room.isFull()` function won't update for them correctly, meaning that the room has too many users. @alankbi, I'm not sure how our backend handles requests like these, will they be done in sequence such that this won't happen or is this bug possible?

If it's possible (that's a big if, I'm not sure), my suggestion would be to fix by checking on the frontend that if a room has more people than the current limit, then the last user to join (presumably the last person in the list, though I don't know if that list preserves order of entry) gets kicked out of the room. This could cause a slight delay where the user joins and then is kicked, but that's a better solution than an accidental violation of the room limit.

As for actions on this PR though, I don't think there are any as this would deal with more complexity and potentially a frontend issue - just noting this down as a potential issue for the future. 👍

Reply:
For reference, I just tested and my prediction was correct - if two users join at the same time, it still adds both of them. (See below image, sorry about the misspelled name lol.)

**Five Users in Four-Limit Room**
<img width="623" alt="Five Users in Four-Limit Room" src="https://user-images.githubusercontent.com/7987279/112285881-01c2a700-8c48-11eb-9e1d-9e93cdb3f6cb.png">

Reply:
Hm yeah that's definitely something that should be fixed in the future - right now it's not too bad cause we don't have any super critical stuff but as we add more things that could cause some bugs. 

In terms of a fix, I think that the best solution might be to implement some kind of [locking system](https://www.baeldung.com/java-jpa-transaction-locks) which addresses the root of the problem (and hopefully would make sure there are zero edge cases/bugs that come up)

Reply:
That's a good idea, I like it 👍 Yeah definitely a future implementation though, makes sense